### ISSUE TITLE
Show bad formatting  on invalid documents

### ISSUE THREAD

**Issue by badsyntax** [badsyntax/vscode-spotless-gradle#156]
I'd like to be able to show the invalid formatting within the document, similar to how other linters work in vscode, for example (TypeScript, eslint & Prettier):

![Screenshot 2020-10-12 at 15 31 15](https://user-images.githubusercontent.com/102141/95758624-9dff7f80-0ca0-11eb-9d8d-0cece25dbe0e.png)

This feature depends on:

- can i run spotlessCheck on individual files? can i use the same stdin/stdout mechanism?
- can i parse the output of spotlessCheck in such a way as to provide useful feedback?
  - i'd like to highlight the diff of the changes similar to the screenshot above
  - ideal data format would be: `line, column, endLine, endColumn`

I'll need to do this through a [vscode language server](https://code.visualstudio.com/api/language-extensions/language-server-extension-guide). 

Typical spotlessCheck output for reference:

```
> The following files had format violations:
      gradle-server/src/main/java/com/github/badsyntax/gradle/GradleBuildCancellation.java
          @@ -25,14 +25,12 @@
           ··}
           
           ··public·static·void·cancelBuild(String·cancellationKey)·throws·GradleCancellationException·{
          -····if·(··Strings.isNullOrEmpty(·cancellationKey))
          -········{
          -························throw·new·GradleCancellationException("No·cancellation·key·specified");
          +····if·(Strings.isNullOrEmpty(cancellationKey))·{
          +······throw·new·GradleCancellationException("No·cancellation·key·specified");
           ····}
           ····CancellationTokenSource·cancellationTokenSource·=·tokens.get(cancellationKey);
           ····if·(cancellationTokenSource·==·null)·{
           
          -
           ······throw·new·GradleCancellationException("Build·is·not·running·for·key:·"·+·cancellationKey);
           ····}·else·{
           
      gradle-server/src/main/java/com/github/badsyntax/gradle/GradleWrapper.java
          @@ -23,7 +23,7 @@
           ······if·(args.length·==·0)·{
           ········throw·new·GradleWrapperException("No·wrapper·args·supplied");
           ······}
          -······················Process·process·=·new·Process(projectRoot);
          +······Process·process·=·new·Process(projectRoot);
           ······process.setUnixCommand(GRADLE_WRAPPER_UNIX);
           ······process.setWindowsCommand(GRADLE_WRAPPER_WINDOWS);
           ······process.exec(args);
  Run './gradlew :gradle-server:spotlessApply' to fix these violations.
```

cc @nedtwigg

**Comment by nedtwigg** [comment]
> can i run spotlessCheck on individual files? can i use the same stdin/stdout mechanism?

No, the `-PspotlessIdeHook` mechanism only works with `spotlessApply`, it disables `spotlessCheck`. 

> can i parse the output of spotlessCheck

The output of `spotlessCheck` is likely to change in the future.

I think it would be easier (for you, not just for Spotless) to use the existing hook.

Spotless has three tasks
- `spotless` - computes the correct content
- `spotlessApply` - copies that content over the existing content
- `spotlessCheck` - throws an error if the existing content is different than the correct content

With the `-PspotlessIdeHook` that you're using already, you get direct access to the correct content, without the disk getting in the way.  If you want, you can easily diff that result against the current content of the buffer.  If the diff is empty, then `spotlessCheck` would succeed.  If it isn't empty, then `spotlessCheck` would fail, and you already have the diff you'd need to create underline marks, but without the confusion of parsing a diff.

**Comment by badsyntax** [comment]
Awesome thanks Ned. I will go with the approach of using `spotlessApply` & `-PspotlessIdeHook` to generate a diff myself, it looks like this should work 👍 

FYI, i consider this feature very experimental. It's unclear if this will work well within in the editor (i'm thinking about possible bad performance), but i'll build a proof of concept to try it out.

**Comment by badsyntax** [comment]
Dev note: Refer to https://github.com/prettier/prettier-linter-helpers/blob/master/index.js to see an approach for generating the text diff suitable for the vscode LS.

**Comment by badsyntax** [comment]
Very early proof of concept in PR https://github.com/badsyntax/vscode-spotless-gradle/pull/160 using the same diffing algorithm as prettier. As expected I'm seeing performance issues and need to decide the best approach for dealing with this.

Series of events:

```
type h
type e
type l
type l
type o
type (space)
type w
type o
type r
type l
type d
```

This spawns many java threads as operations don't complete before running the next `spotlessApply` task. 

The obvious improvement is to debounce the events (don't run spotless for every keypress) and cancel existing spotless tasks before starting a new one.

**Comment by badsyntax** [comment]
It appears as through the extra threads is Gradle spawning new daemon/wrapper processes to parallelise task execution.  This means new tasks are being run before previous tasks have finished.

Ideally there should only be one task running in this scenario and this is what i'm aiming to achieve. There's a bit of async complexity to deal with to get this right.

```
1. type h
2. type e
```

Step 2 should [cancel](https://docs.gradle.org/current/javadoc/org/gradle/tooling/CancellationTokenSource.html#cancel--) the `spotlessApply` task from step 1 if it's still running, and wait for the task to be cancelled before starting another `spotlessApply` task.

Cancelling a task only conceptually requests the build to be cancelled, so we can't "await cancelTask()".

In addition to only running 1 task at a time, to prevent queuing of async operations, we need to store a pointer to the latest request. Then when getting a diff, we can first cancel the existing task, then check `if currentRequest != latestRequest` to prevent queuing new task runs. (A user can type faster than it takes for spotlessApply to complete)

From a code point of view, some initial ideas of how this might be achieved:

```
getDiff(requestId) { 
  this.latestRequest = requestId;
  if (this.runningTask) {
     await this.cancel();  // request the task to be cancelled
     await this.runningTask;  // wait for the task to be cancelled
  }
  // while attempting to cancel the previous task a new request has been sent which invalidates this request
  if (requestId == this.latestRequest) {
    this.runningTask = runSpotlessApply();
    return this.runningTask.then((output) => {
       this.runningTask = null;
       // task could be cancelled! (TODO: check if an error is thrown if task is cancelled)
       if (output != cancelled) {
          return diff(output);
        } else {
          return nodiff;
        }
      }
    })
  } else {
   return noDiff;
  } 
}
```

I can't say i'm 100% confident that i'm following all the async operations correctly, so the above logic might not work, but is a good first proposal that i'm going to try build.

**Comment by nedtwigg** [comment]
Spotless is *much* faster (like 10x) after the JIT has warmed up, and cancelled builds thrash the daemon.  If you limit yourself to only one concurrent formatting task, then Gradle will keep reusing that one single daemon, and it will get faster and faster.

One way to model this is to debounce on both keypresses and the gradle task completion.  Only start a new build if the old one has finished && the last keypress was at least `DEBOUNCE_MS` ago.

RxJava is really useful for modeling ideas like this, dunno how heavy the RxJS port is.

**Comment by badsyntax** [comment]
Yep, it seems like spotless is actually super fast if executed correctly. 

Debouncing the events & single task execution are the main requirements to getting this to work correctly it seems.

One minor thing to clarify:

> Only start a new build if the old one has finished

To prevent showing the wrong state in the editor, i need to run a build using the last state of the text document contents, which is why i can't wait for the previous build to complete before starting a new one. 
I have to instead cancel the existing task, wait for it to finish cancelling, then start a new task with the last document contents.

**Comment by nedtwigg** [comment]
> To prevent showing the wrong state in the editor, i need to run a build using the last state of the text document contents, which is why i can't wait for the previous build to complete before starting a new one.

Just because Spotless gave you a result doesn't mean you have to use it for anything.  I might be wrong, but I bet you can run (100x success) builds faster than you can run (50x success + 50x cancel) builds.  The actual execution of the spotless task is tiny, the slowness is all the project setup.  If the build always succeeds, then the JIT can inline the fast path.  If it gets cancelled even once, then the JIT has to deoptimize.  Not a big deal, but not as fast as it could be.

**Comment by badsyntax** [comment]
Thanks for these tips 👍  Based on what you've explained, i've decided to not cancel at all. I've tried your suggested approach and it seems to work well. The async logic is super simple: "Only start a new build if the old one has finished". 

This does sometimes result in a stale diagnostic state in the editor, but it doesn't appear to be as big of an issue as I originally thought, as most of the time it's just one or two characters off. Once spotless is warmed up, it's hardly an issue at all, so i'm happy with this approach. 

I've also added some additional logic to show diagnostics after opening a document, once the gradle client becomes ready. 

Next I want to provide code actions of the diff for providing quick fixes. And then i'll write some tests for all of these features. I've updated https://github.com/badsyntax/vscode-spotless-gradle/pull/160 with the latest changes.

**Comment by nedtwigg** [comment]
> This does sometimes result in a stale diagnostic state in the editor

I think you can fix this pretty easily.

- Declare a variable `boolean spotlessIsStale`
- When you launch `spotless`, set `spotlessIsStale = false`
- Every time the document changes, set `spotlessIsStale = true` and `clearSquiggles()`
- When a `spotless` result arrives...
  - if `!spotlessIsStale`
    - `drawSquiggles`
  - else
    - ignore result and `launch spotless`
    - (which will set `spotlessIsStale = false`, so that it will `drawSquiggles` the next time the result arrives, assuming the document hasn't changed in the meantime)

**Comment by badsyntax** [comment]
I've implemented something similar to your suggestion and it appears to work quite well. It's looking good. The following screencast shows how it behaviours without any event debouncing on my (fast) machine. With the debouncing it should be even more reliable (but slightly more delayed in giving feedback).

![test](https://user-images.githubusercontent.com/102141/96119456-edce8880-0ee4-11eb-9756-ee533b483316.gif)

**Comment by nedtwigg** [comment]
> Debouncing will reduce the amount of times we call spotless though

If it were my project, I would probably build and merge it without debouncing first (simple always wins!), then open an enhancement issue "Reduce CPU consumption by debouncing Spotless autoformat".  Until CPU usage becomes an issue, why spend time on it :)

**Comment by badsyntax** [comment]
Agreed! 

I'm gonna be merging/releasing #160 soon. It was quite a bit of work but appears to be good enough to release now.

**Comment by badsyntax** [comment]
Released with [1.1.0](https://github.com/badsyntax/vscode-spotless-gradle/releases/tag/1.1.0)
